src/vm/crossloaderallocatorhash.inl

```diff
@@ -0,0 +1,1196 @@
+/*static*/ void DefaultCrossLoaderAllocatorHashTraits<TKey_, TValue_>::DeleteValueInHeapMemory(OBJECTREF keyValueStore, const TValue& value)
+{
+    CONTRACTL
+    {
+        NOTHROW;
+        GC_NOTRIGGER;
+        MODE_COOPERATIVE;
+    }
+    CONTRACTL_END;
+
+    // TODO: Consider optimizing this by changing the add to ensure that the 
+    // values list is sorted, and then doing a binary search for the value instead
+    // of the linear search
+
+    DWORD entriesInArrayTotal;
+    DWORD usedEntries = NoRemoveDefaultCrossLoaderAllocatorHashTraits<TKey,TValue>::ComputeUsedEntries(keyValueStore, &entriesInArrayTotal);
+    TValue* pStartOfValuesData = (TValue*)(((I1ARRAYREF)keyValueStore)->GetDirectPointerToNonObjectElements() + sizeof(TKey));
+
+    for (DWORD iEntry = 0; iEntry < usedEntries; iEntry++)
+    {
+        if (pStartOfValuesData[iEntry] == value)
+        {
+            memmove(pStartOfValuesData + iEntry, pStartOfValuesData + iEntry + 1, (usedEntries - iEntry - 1) * sizeof(TValue));
+            SetUsedEntries(pStartOfValuesData, entriesInArrayTotal, usedEntries - 1);
+            return;
+        }
+    }
+}
+#endif //!DACCESS_COMPILE
+
+/*static*/ inline INT32 GCHeapHashDependentHashTrackerHashTraits::Hash(PtrTypeKey *pValue)
+{
+    LIMITED_METHOD_CONTRACT;
+    return (INT32)*pValue;
+}
+
+/*static*/ inline INT32 GCHeapHashDependentHashTrackerHashTraits::Hash(PTRARRAYREF arr, INT32 index)
+{
+    CONTRACTL
+    {
+        NOTHROW;
+        GC_NOTRIGGER;
+        MODE_COOPERATIVE;
+    }
+    CONTRACTL_END;
+
+    LAHASHDEPENDENTHASHTRACKERREF value = (LAHASHDEPENDENTHASHTRACKERREF)arr->GetAt(index);
+    LoaderAllocator *pLoaderAllocator = value->GetLoaderAllocatorUnsafe();
+    return Hash(&pLoaderAllocator);
+}
+
+/*static*/ inline bool GCHeapHashDependentHashTrackerHashTraits::DoesEntryMatchKey(PTRARRAYREF arr, INT32 index, PtrTypeKey *pKey)
+{
+    CONTRACTL
+    {
+        NOTHROW;
+        GC_NOTRIGGER;
+        MODE_COOPERATIVE;
+    }
+    CONTRACTL_END;
+
+    LAHASHDEPENDENTHASHTRACKERREF value = (LAHASHDEPENDENTHASHTRACKERREF)arr->GetAt(index);
+
+    return value->IsTrackerFor(*pKey);
+}
+
+/*static*/ inline bool GCHeapHashDependentHashTrackerHashTraits::IsDeleted(PTRARRAYREF arr, INT32 index, GCHEAPHASHOBJECTREF gcHeap)
+{
+    CONTRACTL
+    {
+        NOTHROW;
+        GC_NOTRIGGER;
+        MODE_COOPERATIVE;
+    }
+    CONTRACTL_END;
+
+    OBJECTREF valueInHeap = arr->GetAt(index);
+
+    if (valueInHeap == NULL)
+        return false;
+
+    if (gcHeap == valueInHeap)
+        return true;
+
+    // This is a tricky bit of logic used which detects freed loader allocators lazily
+    // and deletes them from the GCHeapHash while looking up or otherwise walking the hashtable
+    // for any purpose.
+    LAHASHDEPENDENTHASHTRACKERREF value = (LAHASHDEPENDENTHASHTRACKERREF)valueInHeap;
+    if (!value->IsLoaderAllocatorLive())
+    {
+#ifndef DACCESS_COMPILE
+        arr->SetAt(index, gcHeap);
```

Comment:
It would be nice to add a comment here that the gcHeap stored in the entry represents a freed entry. I know it is mentioned somewhere else, but it would improve the readability.